constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,534 @@
+
+    # Enforce that values in op_bits columns must be binary.
+    enf is_binary([b]) for b in op_bits
+
+    # Enforce that when the value in is_span column is set to 1, control flow operations cannot be 
+    # executed on the VM, but when is_span flag is 0, only control flow operations can be executed
+    # on the VM.
+    enf 1 - is_span - f_ctrl = 0
+
+
+# Enforces constraint for computing block hashes.
+#
+# Constraint degree: 8
+ev block_hash_computation_constraints([addr, op_bits[7], hasher[8]], [extra, p[4]]):
+    # Operation label for HASHER_LINER_HASH operation.
+    let op_linhash = 3
+
+    # Operation label for HASHER_RETURN_HASH operation.
+    let op_hout = 1
+
+    # Label specifying that we are starting a new hash computation.
+    let m_bp = get_transition_label(op_linhash)
+
+    # Label specifying that we are absorbing the next sequence of 8 elements into an ongoing hash 
+    # computation.
+    let m_abp = get_transition_label(op_linhash)
+
+    # Label specifying that we are reading the result of a hash computation.
+    let m_hout = get_transition_label(op_hout)
+
+    # `alpha` is inherited random values array.
+    let rate_sum = sum([alpha[i + 8] * hasher[i] for i in 0..8])
+    let digest_sum = sum([alpha[i + 8] * hasher[i] for i in 0..4])
+
+    # Variable for initiating a hasher with address addr' and absorbing 8 elements from the hasher
+    # state (hasher[0], ..., hasher[7]) into it.
+    let h_init = alpha[0] + alpha[1] * m_bp + alpha[2] * addr' + rate_sum
+
+    # Variable for the absorption.
+    let h_abp = alpha[0] + alpha[1] * m_abp + alpha[2] * addr' + rate_sum
+
+    # Variable for the result
+    let h_res = alpha[0] + alpha[1] * m_hout + alpha[2] * (addr + 7) + digest_sum
+
+    # Opcode value of the opcode being executed on the virtual machine.
+    let opcode_value = sum([op_bits[i] * 2^i for i in 0..7])
+    
+    # When a control block initializer operation (JOIN, SPLIT, LOOP, CALL, SYSCALL) is executed, a 
+    # new hasher is initialized and the contents of hasher[0], ..., hasher[7] are absorbed into the
+    # hasher. 
+    let u_ctrli = f_ctrli(op_bits, extra) * (h_init + alpha[5] * opcode_value)
+
+    # When SPAN operation is executed, a new hasher is initialized and contents of 
+    # hasher[0], ..., hasher[7] are absorbed into the hasher.
+    let u_span = f_span(op_bits) * h_init
+
+    # When RESPAN operation is executed, contents of hasher[0], ..., hasher[7] (which contain the 
+    # new operation batch) are absorbed into the hasher.
+    let u_respan = f_respan(op_bits, extra) * h_abp
+
+    # When END operation is executed, the hash result is copied into registers 
+    # hasher[0], ..., hasher[3].
+    let u_end = f_end(op_bits, extra) * h_res
+
+    # Enforce the block hash computation constraint.
+    enf p[0]' * (u_ctrli + u_span + u_respan + u_end + 1 - (f_ctrli(op_bits, extra) + 
+        f_span(op_bits) + f_respan(op_bits, extra) + f_end(op_bits, extra))) = p[0]
```

Comment:
there are duplicate computations happening here for `f_ctrli`, `f_span`, `f_respan`, and `f_end`. Instead, it would be good to define variables for these above the respective `u_*` values and then use the values here instead of recomputing